impeller/renderer/backend/metal/gpu_tracer_mtl.mm

```diff
@@ -0,0 +1,53 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <Metal/Metal.h>
+#include "fml/trace_event.h"
+#include "impeller/renderer/backend/metal/context_mtl.h"
+#include "impeller/renderer/backend/metal/formats_mtl.h"
+
+#include <memory>
+
+#include "impeller/renderer/backend/metal/gpu_tracer_mtl.h"
+
+namespace impeller {
+
+void GPUTracerMTL::MarkFrameEnd() {
+#ifdef IMPELLER_DEBUG
+  Lock lock(trace_state_mutex_);
+  current_state_ = (current_state_ + 1) % 16;
+#endif  // IMPELLER_DEBUG
+}
+
+void GPUTracerMTL::RecordCmdBuffer(id<MTLCommandBuffer> buffer) {
+#ifdef IMPELLER_DEBUG
+  Lock lock(trace_state_mutex_);
+  auto current_state = current_state_;
+  trace_states_[current_state].pending_buffers += 1;
+
+  if (@available(ios 10.3, tvos 10.2, macos 10.15, macCatalyst 13.0, *)) {
+    [buffer addCompletedHandler:^(id<MTLCommandBuffer> buffer) {
```

Comment:
I think this has some issues with GPUTracerMTL being deleted (or retained), but I'm not quite sure how to solve that with object-c / c++ interop.

Reply:
Fixed, thanks @gaaclarke !